### ISSUE TITLE
Flutter web : Event listener for opened window

### ISSUE THREAD

**Issue by moioo91120** [flutter/flutter#69040]
Hello,

## Use case

On Flutter web, I would like to login with Twitter.
To do that, I need to access a verify page and check if the url was changed in order to catch the oauth_verifier variable.

On Flutter web, I can open a new web page with 
`final html.WindowBase popupLogin = html.window.open(
        url,
        "popup_name",
        "menubar=no, status=no, scrollbars=no, menubar=no, width=1000, height=500");`

Then I would like to add an event listener to catch when the url change :
`popupLogin.addEventListener("hashchange", (event) {
      print(event);
    });`

However, I get the following error : You can only attach EventListeners to your own window

## Proposal

Allow user to add event listener on other windows, like what we can do with standard JS.

Thank you ! 
Best regards,
Fabien

**Comment by Awaisoop** [comment]
can you please share full code here

**Comment by moioo91120** [comment]
Hello, it would be something like that.

The event part is not fully working as I would like to print it to get the correct path.

` _signInWithWeb() async {
    final http.Response response = await this.api.get(path + "/connect");
    final Oauth1Token token = Oauth1Token.fromJson(json.decode(response.body));

    // await launch(TW_VERIFIER + token.token, forceWebView: true);
    final html.WindowBase popupLogin = html.window.open(
        TW_VERIFIER + token.token,
        "Twitter",
        "menubar=no, status=no, scrollbars=no, menubar=no, width=1000, height=500");

    popupLogin.addEventListener("hashchange", (event) => {
      print(event),
      if (event.path.contains("oauth_verifier=")) {
         // get the verifier token here
         
        popupLogin.close()
      }
    });

  }`

Thanks,
Fabien

**Comment by Awaisoop** [comment]
ok first tell me that you are using this on hosting or debug mode ?

**Comment by moioo91120** [comment]
For now, on debug mode.

When it's working, it's going to be hosted on the web (CDN + AWS S3).

**Comment by Zhang-Jony** [comment]
@Awaisoop any workaround？

**Comment by sigmundch** [comment]
@yjbanov - correct, however if I recall correctly part of the issue with cross-frames is that each frame in JavaScript has their separate isolate and as a result the way dart2js associates type information can break. @rakudrama - do you recall if that was the reason for limiting the exposure of those APIs?

/cc @srujzs


I believe you may be able to workaround this for now using JS-interop and directly calling the event listener using `js_util`:
```dart
import 'dart:js_util' as js_util;

...
   js_util.callMethod("addEventListener", ["hashchange", allowInterop((event) { print(event); })]);
```

**Comment by rakudrama** [comment]
dart2js should be able to dispatch correctly on same-origin cross-frame objects; the dispatch info won't be pre-installed like on the main DOM classes.
A reason to separate out the cross-frame window is to make cross-frame access go through a single point to allow dispatch info to be installed for Dartium and DDC.

**Comment by henriklagergren** [comment]
Any updates on this issue?

**Comment by diezep** [comment]
Any updates?

**Comment by sigmundch** [comment]
Sorry, not much has changed on this issue yet. Did the workaround in the comment above work for you?

**Comment by moioo91120** [comment]
Hello,

So I tried 

```
final html.WindowBase popupLogin = html.window.open(
          TW_VERIFIER + token.token,
          "Twitter",
          "menubar=no, status=no, scrollbars=no, menubar=no, width=1000, height=500");
      js_util.callMethod(popupLogin, "addEventListener", [
        "hashchange",
        allowInterop((event) {
          print(event);
        })
      ]);

      popupLogin.close();
```

Still got the error 
UnsupportedError (Unsupported operation: You can only attach EventListeners to your own window.)

**Comment by moioo91120** [comment]
Hello, any news ? can we expect this in any futur version ?

**Comment by rakudrama** [comment]
I believe this did not work because the value returned by `html.window.open` is not a browser window, but a `_DOMWindowCrossFrame` wrapper (a Dart object with the unsupported operations).

You might try using `js_util` to create the window directly
```
Object? popupLogin = js_util.callMethod(html.window, "open", ...);
```
Note that I used `Object?` as the type since the other window does not respond correctly to `is Window` consistently across browsers and development / production (this is why we have the cross frame window). You can only use js_util on this version.

/cc @srujzs - something to consider for the current html design work.

**Comment by srujzs** [comment]
As an aside, might be worth mentioning this issue has had some discussion years ago here: #17936.

**Comment by moioo91120** [comment]
Ok, thanks for the answer @rakudrama .
So my next try was 


    Object popupLogin = js_util.callMethod(html.window, "open", [
      urlAuthorizer,
    ]); 

    js_util.callMethod(popupLogin, "addEventListener", [
      "locationchange",
      allowInterop((event) {
        print("addEventListener locationchange");
        print(event);
        callback();
      })
    ]);
    js_util.callMethod(popupLogin, "addEventListener", [
      "hashchange",
      allowInterop((event) {
        print("addEventListener hashchange");
        print(event);
        callback();
      })
    ]);

I added event on locationchange and hashchange in hope that 1 of them would be triggered.

Resultats are 
- No more error
- But actually no print, callback or anything from the addEventListener

I also tried to cast the result as a Window so that I could use the addEventListener method, but there was an error saying that the returned object was a NativeJavaScriptObject and not a Window

**Comment by moioo91120** [comment]
![image](https://user-images.githubusercontent.com/15015669/111759546-00872800-889e-11eb-8fab-5d851bbab22e.png)
Actually in the new openned tab, I don't see the event listener, so I'm not sure the js_util.callMethod(..., "addEventListener, ...) addded the events.

**Comment by srujzs** [comment]
Maybe there might be some same-origin policy issues with your opened window? Syntactically, it works for me using `about:blank` as the opened window instead and a click to simulate an event:

```
import 'dart:html' as html;
import 'package:js/js.dart';
import 'package:js/js_util.dart' as js_util;

void main() {
  var popupLogin = js_util.callMethod(html.window, "open", [
    "about:blank",
  ]);

  js_util.callMethod(popupLogin, "addEventListener", [
    "click",
    allowInterop((event) {
      print("addEventListener click");
    })
  ]);
}
```

**Comment by moioo91120** [comment]
Hello,

Thank you.
Indeed it works for "about:blank", but not for a external url , ex 



    Object popupLogin =
        js_util.callMethod(html.window, "open", ["https://www.google.com"]);
    js_util.callMethod(popupLogin, "addEventListener", [
      "click",
      allowInterop((event) {
        print("addEventListener locationchange");
        print(event);
        callback();
      })
    ]);

We have no click event triggered

**Comment by unipenlabs** [comment]
Hello, any update on this? We need it :) Thanks!

**Comment by srujzs** [comment]
Hi,

I think this issue is complicated to resolve within `dart:html`, so no updates yet, sorry. Does the above workaround work for your purposes? I believe the reason it works for `about:blank` but not an external url is because of the browser's same-origin policy and not specific to `dart:html` unless I'm missing something.

**Comment by moioo91120** [comment]
Hi @srujzs , it did not solve my problem. I ended opening a new tab and saving a cookie at the end of my Twitter auth , with some kind of loop in original window to get the data.

Pretty ugly but at least it works

**Comment by luisfelipesd** [comment]
> Hi @srujzs , it did not solve my problem. I ended opening a new tab and saving a cookie at the end of my Twitter auth , with some kind of loop in original window to get the data.
> 
> Pretty ugly but at least it works

Can you share your code?

**Comment by moioo91120** [comment]
@luisfelipesd  nothing fancy 

I open windows on twitter oauth workflow  and launch an infinite loop with sleep until I get a cookie

    final http.Response response = await this.api.get(path + "/connect");
    Oauth1Token token = Oauth1Token.fromJson(json.decode(response.body));
    final String urlAuthorizer = TW_VERIFIER + token.token;
    html.window.localStorage.remove(PREF_NAME_VERIFIER);

    html.window.open(urlAuthorizer, "Twitter",
        "menubar=no, status=no, scrollbars=no, menubar=no, width=1000, height=500");

    final int nbTry = 100;
    for (int i = 0; i < nbTry; i++) {
      await Future.delayed(Duration(seconds: 1));
      if (html.window.localStorage[PREF_NAME_VERIFIER] != null) {
        token = token.copyWith(
            tokenVerifier: html.window.localStorage[PREF_NAME_VERIFIER]);
        html.window.localStorage.remove(PREF_NAME_VERIFIER);
        break;
      }
      i++;
    }


Then at some point, the twitter worklow get redirected to my app and I just set my cookie 

    html.window.localStorage[TwitterProvider.PREF_NAME_VERIFIER] =
        tokenVerifier;

**Comment by wizard11** [comment]
Any updates on the clean approach?

**Comment by srujzs** [comment]
No updates within `dart:html` for this yet, sorry!

**Comment by ceboma** [comment]
Hi, try to use this package : universal_html

Example  : 

import 'dart:convert';
import 'package:universal_html/html.dart' as html;

printDoc(fileName, base64String) {
    final bytes = base64.decode(base64String) ;
    final blob = html.Blob([bytes], 'application/pdf');
    final url = html.Url.createObjectUrlFromBlob(blob);
    final window = html.window.open(url, fileName + '.pdf');
}

**Comment by Moumene** [comment]
I found an easy way to do so with [flutter_web_auth](https://pub.dev/packages/flutter_web_auth) 

``` dart
// Present the dialog to the user
final result = await FlutterWebAuth.authenticate(url: url.toString(), callbackUrlScheme: callbackUrlScheme);

// Extract code from resulting url
final code = Uri.parse(result).queryParameters['code'];
```

**Comment by olion500** [comment]
I found an another way to do with redirect_url

https://github.com/TesteurManiak/flutter_web_twitch_auth

**Comment by ljmatan** [comment]
Does anything work with this framework?

**Comment by niteshsh4rma** [comment]
https://github.com/RasmusSlothJensen/flutter_oauth_external_window This might help

**Comment by yehya-qassim** [comment]
You can check the solution i created for such case: [the issue and solution](https://stackoverflow.com/questions/73406343/flutter-web-event-listener-on-other-windows/73408566#73408566)

**Comment by shravan-g-k** [comment]
Is there any solution yet for this problem

**Comment by shravan-g-k** [comment]
I tried your solution @yehya-qassim but the event listener never gets the message as the window that it focuses is the main window and not the popup window

**Comment by fennel-yehya-qassim** [comment]
@shravan-g-k that's true since you are redirecting the same window to the callback url which is better since many browsers prevent popups, I actually changed my way to the same way you are doing.

In that case what you need to do is that navigate to the callback.html page then inside it extract the code from the url then navigate back to the back view such as:

`var code = window.location.href.split("code=")[1].split("&")[0];`
`window.location.href = url + "/#some_router_name_you_are_using?code=" + code;`
`window.location.href = url + "/#login";`

Then inside main.dart check if the url have code if it does extract it and use it to get token authenticate the user and navigate to the main app instead of login.

Let me know if you need help!

**Comment by yehya-qassim** [comment]
@shravan-g-k  that's true since you are redirecting the same window to the callback url which is better since many browsers prevent popups, I actually changed my way to the same way you are doing.

In that case what you need to do is that navigate to the callback.html page then inside it extract the code from the url then navigate back to the back view such as:
`var code = window.location.href.split("code=")[1].split("&")[0];`
 `window.location.href = url + "/#some_router_name_you_are_using?code=" + code; `
 `window.location.href = url + "/#login";`
 
Then inside main.dart check if the url have code if it does extract it and use it to get token authenticate the user and navigate to the main app instead of login.

Let me know if you need help!

**Comment by shravan-g-k** [comment]
@yehya-qassim  Thank You so much your solution works beautifully .

**Comment by shravan-g-k** [comment]
I just have a doubt it works when the redirct uri is callback.html but does't when its any other flutter page

**Comment by yehya-qassim** [comment]
> I just have a doubt it works when the redirct uri is callback.html but does't when its any other flutter page

It has to be callback.html you can use it only for authentication, it doesn't have any other use.
but if you meant you have another page not named callback.html you can do the following

`var url = window.location.href.split('/your_page_name.html')[0]`

**Comment by shravan-g-k** [comment]
Ok now I get it . Thank You

**Comment by nshahan** [comment]
@srujzs
There is a long thread here and I don't think we have anything to say about `dart:html`, but do you know if we have any other updates to share since the rollout of `packae:web`?
